src/access/tests/TestAccessControl.cpp

```diff
@@ -28,497 +29,980 @@ namespace {
-        kCluster    = 1 << 2,
-    };
+        memset(this, 0, sizeof(*this));
+    }
+
+    bool IsEmpty() const
+    {
+        return authMode == AuthMode::kNone;
+    }
+
+    size_t GetSubjectCount() const
+    {
+        size_t count = 0;
+        for (auto & subject : subjects)
+        {
+            if (subject == kUndefinedNodeId)
+            {
+                break;
+            }
+            count++;
+        }
+        return count;
+    }
+
+    void AddSubject(size_t * index, NodeId subject)
+    {
+        size_t count = GetSubjectCount();
+        if (count < kMaxSubjects)
+        {
+            subjects[count] = subject;
+            if (index)
+            {
+                *index = count;
+            }
+        }
+    }
 
-    int flags = kEmptyFlags;
-    DeviceTypeId deviceType;
-    EndpointId endpoint;
-    ClusterId cluster;
+    void RemoveSubject(size_t index)
+    {
+        size_t count = GetSubjectCount();
+        if (index < count)
+        {
+            while (++index < kMaxSubjects)
+            {
+                subjects[index - 1] = subjects[index];
+            }
+            subjects[kMaxSubjects - 1] = { 0 };
+        }
+    }
+
+    size_t GetTargetCount() const
+    {
+        size_t count = 0;
+        for (auto & target : targets)
+        {
+            if (target.flags == 0)
+            {
+                break;
+            }
+            count++;
+        }
+        return count;
+    }
+
+    void AddTarget(size_t * index, const Target & target)
+    {
+        size_t count = GetTargetCount();
+        if (count < kMaxTargets)
+        {
+            targets[count] = target;
+            if (index)
+            {
+                *index = count;
+            }
+        }
+    }
+
+    void RemoveTarget(size_t index)
+    {
+        size_t count = GetTargetCount();
+        if (index < count)
+        {
+            while (++index < kMaxTargets)
+            {
+                targets[index - 1] = targets[index];
+            }
+            targets[kMaxTargets - 1] = { 0 };
+        }
+    }
 };
 
-TestTarget Target(EndpointId endpoint, ClusterId cluster)
+CHIP_ERROR CompareEntry(const Entry & entry, const EntryData & entryData)
+{
+    AuthMode authMode;
+    ReturnErrorOnFailure(entry.GetAuthMode(authMode));
+    ReturnErrorCodeIf(authMode != entryData.authMode, CHIP_ERROR_INCORRECT_STATE);
+    FabricIndex fabricIndex;
+    ReturnErrorOnFailure(entry.GetFabricIndex(fabricIndex));
+    ReturnErrorCodeIf(fabricIndex != entryData.fabricIndex, CHIP_ERROR_INCORRECT_STATE);
+    Privilege privilege;
+    ReturnErrorOnFailure(entry.GetPrivilege(privilege));
+    ReturnErrorCodeIf(privilege != entryData.privilege, CHIP_ERROR_INCORRECT_STATE);
+    size_t subjectCount;
+    ReturnErrorOnFailure(entry.GetSubjectCount(subjectCount));
+    ReturnErrorCodeIf(subjectCount != entryData.GetSubjectCount(), CHIP_ERROR_INCORRECT_STATE);
+    for (size_t j = 0; j < subjectCount; ++j)
```

Comment:
Suggest naming the indices to be self-documenting instead of `j` everywhere: `subjectIdx`, `targetIdx`, etc